aten/src/ATen/native/sparse/cuda/SoftMax.cu

```diff
@@ -0,0 +1,776 @@
+  Tensor result = at::empty_like(input);
+  if (input.numel() == 0) {
+    return result;
+  }
+  TORCH_CHECK(
+      dim_ >= 0 && dim_ < input.dim(),
+      "dim must be non-negative and less than input dimensions");
+
+  AT_DISPATCH_FLOATING_TYPES(input.scalar_type(), "softmax", [&] {
+    cuda_sparse_coo_softmax<scalar_t, false>(result, input, dim_);
+  });
+  return result;
+}
+
+Tensor log_softmax_sparse_cuda(
+    const Tensor& input_,
+    const int64_t dim_,
+    const bool half_to_float) {
+  TORCH_INTERNAL_ASSERT(input_.is_sparse());
+
+  TORCH_CHECK(!half_to_float, "conversion is supported for Half type only");
+  auto input = input_.coalesce();
+  Tensor result = at::empty_like(input);
+  if (input.numel() == 0) {
+    return result;
+  }
+  TORCH_CHECK(
+      dim_ >= 0 && dim_ < input.dim(),
+      "dim must be non-negative and less than input dimensions");
+
+  AT_DISPATCH_FLOATING_TYPES(input.scalar_type(), "softmax", [&] {
+    cuda_sparse_coo_softmax<scalar_t, true>(result, input, dim_);
+  });
+  return result;
+}
+
+Tensor softmax_backward_sparse_cuda(
+    const Tensor& grad_,
+    const Tensor& output_,
+    int64_t dim_,
+    const Tensor& input_) {
+  TensorArg grad_arg{grad_, "grad", 1}, output_arg{output_, "output", 2};
+  checkSameSize("softmax_backward", grad_arg, output_arg);
+
+  int64_t dim = maybe_wrap_dim(dim_, grad_.dim());
+
+  auto grad = grad_.coalesce();
+  auto output = output_.coalesce();
+
+  Tensor grad_input = at::native::empty_like(output);
+  if (output.numel() == 0) {
+    return grad_input;
+  }
+
+  TORCH_CHECK(
+      dim >= 0 && dim < grad.dim(),
+      "dim must be non-negative and less than input dimensions");
+  TORCH_CHECK(
+      grad.sparse_dim() == output.sparse_dim(),
+      "grad and output sparse dimensions must be equal");
+
+  AT_DISPATCH_FLOATING_TYPES(grad.scalar_type(), "softmax_backward", [&] {
+    cuda_sparse_coo_softmax_backward<scalar_t, false>(
+        grad_input, grad, output, dim);
+  });
+  return grad_input;
+}
+
+Tensor log_softmax_backward_sparse_cuda(
+    const Tensor& grad_,
+    const Tensor& output_,
+    int64_t dim_,
+    const Tensor& input_) {
+  TensorArg grad_arg{grad_, "grad", 1}, output_arg{output_, "output", 2};
+  checkSameSize("softmax_backward", grad_arg, output_arg);
+
+  int64_t dim = maybe_wrap_dim(dim_, grad_.dim());
+
+  auto grad = grad_.coalesce();
+  auto output = output_.coalesce();
+
+  Tensor grad_input = at::native::empty_like(output);
+  if (output.numel() == 0) {
+    return grad_input;
+  }
+
+  TORCH_CHECK(
+      dim >= 0 && dim < grad.dim(),
+      "dim must be non-negative and less than input dimensions");
+  TORCH_CHECK(
+      grad.sparse_dim() == output.sparse_dim(),
+      "grad and output sparse dimensions must be equal");
+
+  AT_DISPATCH_FLOATING_TYPES(grad.scalar_type(), "softmax_backward", [&] {
+    cuda_sparse_coo_softmax_backward<scalar_t, true>(
+        grad_input, grad, output, dim);
+  });
+  return grad_input;
+}
```

Comment:
This is repeating the functions in aten/src/ATen/native/sparse/SoftMax.cpp .

To avoid the code replication, I suggest updating the corresponding functions in aten/src/ATen/native/sparse/SoftMax.cpp with the following (example is given only for `softmax_sparse`, updates to other functions are analogous):
```
Tensor softmax_sparse(const Tensor& input_, const int64_t dim_, const bool half_to_float) {
  TORCH_INTERNAL_ASSERT(input_.is_sparse());
  TORCH_CHECK(!half_to_float, "softmax with half to float conversion is not supported on CPU");
  auto input = input_.coalesce();
  Tensor output = at::native::empty_like(input);
  if (input.numel() == 0) {
    return output;
  }
  TORCH_CHECK(dim_ >= 0 && dim_ < input.dim(),
              "dim must be non-negative and less than input dimensions");
  AT_DISPATCH_FLOATING_TYPES(input.scalar_type(), "softmax", [&] {
      if (input.is_cuda()) {
         cuda_sparse_coo_softmax<scalar_t, false>(output, input, dim_);
      } else {
         cpu_sparse_coo_softmax<scalar_t, false>(output, input, dim_);
      }
    });
  return output;
}
```
Here `softmax_sparse` is the original `softmax_sparse_cpu` and the function `softmax_sparse_cuda` in the given file can be deleted.

After this renaming, update the native_functions.yaml accordingly:
```
- func: _sparse_softmax(Tensor self, int dim, bool half_to_float) -> Tensor
  use_c10_dispatcher: full
  dispatch:
    SparseCPU, SparseCUDA: softmax_sparse
```

Reply:
I am going to use the device dispatcher to refactor this part of the code due to the cuda kernel code should live inside a .cu file.